File: day_06/calculate_route.go
package day06

import (
	"advent/internal/input"
	"fmt"
	"strconv"
	"strings"
	"sync"
)

type Direction int64

const (
	north Direction = iota
	east
	south
	west
)

var mutex sync.Mutex

// Calculate the travel in North-South direction. Returns true if the route exist the grid.
// Otherwise false
func (route *guardRoute) travelNorthSouth(labMap map[input.Coordinates]bool,
	loopDetection bool) (int64, *guardRoute) {
	previousPosition := route.position

	for y := route.position.Y; y < route.mapSize.Y; {
		if y < 0 {
			break
		}

		route.position.Y = y
		if labMap[route.position] {
			// Hitting a block
			route.position = previousPosition
			return 0, route
		}

		if loopDetection {
			trailPoint := convertToStrWDirection(route.position, route.direction)
			if route.trail[trailPoint] {
				// fmt.Printf("NS Loop detected %v\n", route.position)
				return 2, route
			}
			route.trail[trailPoint] = true
		} else {
			trailPoint := convertToStr(route.position)
			route.trail[trailPoint] = true
		}
		previousPosition = route.position

		switch route.direction {
		case north:
			y--
		case south:
			y++
		default:
			panic("unexpected direction")
		}

	}
	route.position = previousPosition

	return 1, route // Found the exit
}

// Calculate the travel in East-West direction. Returns true if the route exist the grid.
// Otherwise false
func (route *guardRoute) travelEastWest(labMap map[input.Coordinates]bool,
	loopDetection bool) (int64, *guardRoute) {
	previousPosition := route.position
	for x := route.position.X; x < route.mapSize.X; {
		if x < 0 {
			break
		}
		route.position.X = x

		if labMap[route.position] {
			route.position = previousPosition
			return 0, route
		}

		if loopDetection {
			trailPoint := convertToStrWDirection(route.position, route.direction)
			if route.trail[trailPoint] {
				// fmt.Printf("EW Loop detected %v\n", route.position)
				return 2, route
			}
			route.trail[trailPoint] = true
		} else {
			trailPoint := convertToStr(route.position)
			route.trail[trailPoint] = true
		}
		previousPosition = route.position

		switch route.direction {
		case east:
			x++
		case west:
			x--
		default:
			panic("unexpected direction")
		}
	}

	route.position = previousPosition
	return 1, route // Found the exit
}

// addPading prep ends the ordinal string with zeroes so that the length is always 3
func addPading(orginalString string) string {
	var padded string
	switch len(orginalString) {
	case 1:
		padded = "00" + orginalString
	case 2:
		padded = "0" + orginalString
	case 3:
		padded = orginalString
	default:
		fmt.Printf("Original: %v\n", orginalString)
		panic("Cannot convert to sting with these assumptions")
	}

	return padded
}

// Returns a string representation of the input coordinates augmented with the given direction
func convertToStrWDirection(originalObject input.Coordinates, direction Direction) string {
	var coordinateStr string
	var directionStr string

	if originalObject.X < 0 || originalObject.Y < 0 {
		panic("Encountered illegal coordinates [with directions]")
	}
	xStr := addPading(strconv.Itoa(originalObject.X))
	yStr := addPading(strconv.Itoa(originalObject.Y))
	switch direction {
	case north:
		directionStr = "north"
	case east:
		directionStr = "east"
	case south:
		directionStr = "south"
	case west:
		directionStr = "west"
	}

	coordinateStr = xStr + "," + yStr + "," + directionStr

	return coordinateStr
}

// Returns a string representation of the input coordinates
func convertToStr(originalObject input.Coordinates) string {
	var coordinateStr string

	if originalObject.X < 0 || originalObject.Y < 0 {
		panic("Encountered illegal coordinates []")
	}
	xStr := addPading(strconv.Itoa(originalObject.X))
	yStr := addPading(strconv.Itoa(originalObject.Y))

	coordinateStr = xStr + "," + yStr

	return coordinateStr
}

func convertToCoordinate(coordinateString string) input.Coordinates {
	var converted input.Coordinates

	splitted := strings.Split(coordinateString, ",")

	xInt, err := strconv.Atoi(splitted[0])
	converted.X = xInt
	if err != nil {
		errorStr := fmt.Sprintf("Cannot convert %v", coordinateString)
		panic(errorStr)
	}

	yInt, err := strconv.Atoi(splitted[1])
	converted.Y = yInt
	if err != nil {
		errorStr := fmt.Sprintf("Cannot convert %v", coordinateString)
		panic(errorStr)
	}

	return converted
}

// Calculates the travel to the direction defined in the route structure. Returns updated route
// struck and boolean true if an exit was found. The boolean will be false otherwise
func (route *guardRoute) travel(labMap map[input.Coordinates]bool,
	loopDetection bool) (int64, *guardRoute) {
	debug := false
	var exitFound int64
	switch route.direction {
	case north:
		exitFound, route = route.travelNorthSouth(labMap, loopDetection)
	case east:
		exitFound, route = route.travelEastWest(labMap, loopDetection)
	case south:
		exitFound, route = route.travelNorthSouth(labMap, loopDetection)
	case west:
		exitFound, route = route.travelEastWest(labMap, loopDetection)
	default:
		panic("unexpected direction")
	}

	if debug {
		fmt.Println(labMap)
	}

	return exitFound, route
}

func (route *guardRoute) guardNavigation(blockMap map[input.Coordinates]bool, loopDetection bool) int {
	var exitFound int64
	killSwitch := 1200
	i := 0
	j := 0
	allDirections := [...]Direction{north, east, south, west}

	for true {
		if j > killSwitch {
			panic("The kill switch was triggered")
		}

		route.direction = allDirections[i]
		exitFound, route = route.travel(blockMap, loopDetection)
		if exitFound == 1 {
			return 0
		}
		if exitFound == 2 {
			// fmt.Printf("Loop detected at position %v\n", route.position)
			return 1
		}
		i++
		j++

		if i >= len(allDirections) {
			i = 0
		}

	}

	return 0
}

type guardRoute struct {
	mapSize   input.Coordinates
	position  input.Coordinates
	trail     map[string]bool
	direction Direction
}

func afterAddingNewBlock(wg *sync.WaitGroup, total *int, startingPoint input.Coordinates,
	blockCoordinates map[input.Coordinates]bool, currentAddition string,
	mapSize input.Coordinates) {
	var currenRoute guardRoute
	tryBlock := convertToCoordinate(currentAddition)

	currenRoute.trail = make(map[string]bool)
	currenRoute.position = startingPoint
	currenRoute.direction = north
	currenRoute.mapSize = mapSize

	// fmt.Printf("blockCoordinates %p\n", &blockCoordinates)
	improvedBlock := make(map[input.Coordinates]bool)

	for coordinate, v := range blockCoordinates {
		if v {
			improvedBlock[coordinate] = true
		}
	}

	improvedBlock[tryBlock] = true
	res := currenRoute.guardNavigation(improvedBlock, true)

	if res == 1 {
		// print(res)
		mutex.Lock()
		defer mutex.Unlock()
		*total++
	}

	wg.Done()

}

func CalculateRoute() {
	var wg sync.WaitGroup
	var route guardRoute
	// inputLines := input.GetInputV2("./day_06/test-input-1.txt")
	inputLines := input.GetInputV2("../downloads/input-day6.txt")
	blockCoordinates := input.GetCoordinates(inputLines, "#")
	startingPoint := input.GetCoordinates(inputLines, "^")

	blockMap := make(map[input.Coordinates]bool)
	for _, v := range blockCoordinates {
		blockMap[v] = true
	}

	route.mapSize = input.GetGridSize(inputLines)
	route.position = startingPoint[0]
	route.direction = north
	route.trail = make(map[string]bool)

	fmt.Printf("Found %v blocks in total.\n", len(blockCoordinates))
	fmt.Printf("Found %v guards in total.\n", len(startingPoint))
	fmt.Printf("The total size of the map is %v by %v\n", route.mapSize.X, route.mapSize.Y)

	route.guardNavigation(blockMap, false)
	fmt.Println("-----")
	fmt.Printf("lenght %v\n", len(route.trail))

	// Ignore the starting point
	total := 0
	startCoord := convertToStr(startingPoint[0])
	route.trail[startCoord] = false
	originalRoute := route.trail

	for coordinate, v := range originalRoute {
		if v {
			route.trail = make(map[string]bool)
			route.position = startingPoint[0]
			route.direction = north

			improvedMap := blockMap
			wg.Add(1)
			go afterAddingNewBlock(&wg, &total, startingPoint[0], improvedMap, coordinate,
				input.GetGridSize(inputLines))
		}
	}

	wg.Wait()

	fmt.Printf("The number of possible loops is: %v\n", total)
}

File: day_06/calculate_route.py
import sys


class Coordinates:
    def __init__(self, x, y):
        self.X = x
        self.Y = y

    def add_padding(self, raw_string):
        padded = ""
        string_len = len(raw_string)
        if string_len == 1:
            padded = "00" + raw_string
        elif string_len == 2:
            padded = "0" + raw_string
        elif string_len == 3:
            padded = raw_string
        else:
            print(f"ERRRO: Cannot padd this: {raw_string}")

        return padded

    def get_str(self):
        raw_x = str(self.X)
        raw_y = str(self.Y)

        string_x = self.add_padding(raw_x)
        string_y = self.add_padding(raw_y)

        return string_x + "," + string_y

    def get_str_w_directions(self, direction):
        only_coord = self.get_str()
        return only_coord + "," + direction


def get_coordinates(input_lines, target):
    y = 0

    coordinates = []
    for current_line in input_lines:
        x = 0
        for letter in current_line:
            if letter == target:
                block = Coordinates(x, y)
                coordinates.append(block)
            x += 1
        y += 1

    return coordinates


def get_grid_size(input_lines):
    x = len(input_lines[0])
    y = len(input_lines)

    return Coordinates(x,y)


def get_input(file_name):
    with open(file_name, 'r', encoding='UTF-8') as handle:
        input_lines = handle.readlines()

    return input_lines

class GuardRoute:
    def __init__(self, map_size, position, direction):
        self.map_size = map_size
        self.position = position
        self.trail = {}
        self.direction = direction

    def travel_north_south(self, lab_map, loop_detection):
        previous_position = Coordinates(self.position.X, self.position.Y)
        y = self.position.Y

        while y < self.map_size.Y:
            if y < 0:
                break

            self.position.Y = y

            if lab_map.get((self.position.X, self.position.Y), False):
                self.position = previous_position
                return 0, self

            if loop_detection:
                trail_point = self.position.get_str_w_directions(self.direction)
                if self.trail.get(trail_point):
                    return 2, self
            else:
                trail_point = self.position.get_str()

            self.trail[trail_point] = True
            previous_position = Coordinates(self.position.X, self.position.Y)

            if self.direction == "north":
                y -= 1
            elif self.direction == "south":
                y += 1
            else:
                raise ValueError("unexpected direction")

        self.position = previous_position
        return 1, self

    def travel_east_west(self, lab_map, loop_detection):
        previous_position = Coordinates(self.position.X, self.position.Y)
        x = self.position.X

        while x < self.map_size.X:
            if x < 0:
                break

            self.position.X = x

            if lab_map.get((self.position.X, self.position.Y), False):
                self.position = previous_position
                return 0, self

            if loop_detection:
                trail_point = self.position.get_str_w_directions(self.direction)
                if self.trail.get(trail_point):
                    return 2, self
            else:
                trail_point = self.position.get_str()

            self.trail[trail_point] = True
            previous_position = Coordinates(self.position.X, self.position.Y)

            if self.direction == "east":
                x += 1
            elif self.direction == "west":
                x -= 1
            else:
                raise ValueError("unexpected direction")

        self.position = previous_position
        return 1, self

    def travel(self, lab_map, loop_detection):
        if self.direction == "north" or self.direction == "south":
            return self.travel_north_south(lab_map, loop_detection)
        elif self.direction == "east" or self.direction == "west":
            return self.travel_east_west(lab_map, loop_detection)
        else:
            raise ValueError("unexpected direction")

    def guard_navigation(self, block_map, loop_detection):
        exit_found = 0
        kill_switch = 1200
        i = j = 0
        all_directions = ["north", "east", "south", "west"]

        while True:
            if j > kill_switch:
                raise RuntimeError("The kill switch was triggered")

            self.direction = all_directions[i]
            exit_found, _ = self.travel(block_map, loop_detection)

            if exit_found == 1:
                return 0
            if exit_found == 2:
                return 1

            i = (i + 1) % len(all_directions)
            j += 1

def main():
    print("Hello")

    input = get_input("./day_06/test-input-1.txt")
    block_coordinates = get_coordinates(input, "#")
    starting_point = get_coordinates(input, "^")
    map_size = get_grid_size(input)

    block_map = {}
    for block in block_coordinates:
        block_str = block.get_str()
        block_map[block_str] = True

    route = GuardRoute(map_size=map_size, position=starting_point[0], direction="north")
    route.guard_navigation(block_map, False)
    return 0


if __name__ == "__main__":
    sys.exit(main())
